### ISSUE TITLE
Chained operations with annotated types fails for writing to stream

### ISSUE THREAD

**Issue by eulores** [sealmove/binarylang#22]
In the following test case, only field1 and field2 are correctly encoded, but field3 fails because of the chained operations.

```nim
import binarylang
from sequtils import mapIt
from strutils import join

const broken = false

template seqGet*(parse, parsed, output: untyped) =
  echo "seqGet: ", typeof(parsed), " ==> ", typeof(output)
  when broken:
    parse
    output = parsed

template seqPut*(encode, encoded, output: untyped) =
  echo "seqPut: ", typeof(encoded), " ==> ", typeof(output)
  when broken:
    output = encoded
    encode

template strGet*(parse, parsed, output: untyped) =
  echo "strGet: ", typeof(parsed), " ==> ", typeof(output)
  when broken:
    parse
    output = parsed.mapIt(it.char).join

template strPut*(encode, encoded, output: untyped) =
  echo "strPut: ", typeof(encoded), " ==> ", typeof(output)
  when broken:
    output = encoded.mapIt(uint8(it.ord))
    encode

struct(testCase):
  u8 {seq}: field1[4]
  u8 {str[string]}: field2[4]
  u8 {seq, str[string]}: field3[4]

when isMainModule:
  echo("Test case for binarylang")
  discard TestCase(field1: @[1'u8,2,3,4], field2:"ABCD", field3:"EFGH").fromTestCase
```

Generated output:
```
Test case for binarylang
seqPut: seq[uint8] ==> seq[uint8]
strPut: string ==> seq[uint8]
strPut: string ==> string      # error
seqPut: string ==> seq[uint8]  # error
```

Expected output:
```
Test case for binarylang
seqPut: seq[uint8] ==> seq[uint8]
strPut: string ==> seq[uint8]
strPut: string ==> seq[uint8]
seqPut: seq[uint8] ==> seq[uint8]
```

A quick fix that works for me is changing the index from `i` to `i-1` in proc `generateWriter` in file `binarylang.nim`
```nim
      for i in countdown(f.ops.len - 1, 1):
        var
          # typeImpl = f.ops[i].typ
          typeImpl = f.ops[i-1].typ # here is the patch
          write = quote do: `outputSym` = `encoded`
          op = newCall(ident(f.ops[i].name & "put"), write, wSym, encoded)
```

You have created an amazing and very useful library! There are a number of improvements not yet in the last release. Would you please consider issuing a new release? Thanks!

**Comment by sealmove** [comment]
Hello. I am happy that you find it useful. No promises (limited time) but I will try to dedicate some time for a new release and possibly to make the code more maintainable (split in multiple files maybe).

**Comment by sealmove** [comment]
This is related to https://github.com/sealmove/binarylang/issues/19. Type inference in [Operations](https://sealmove.github.io/binarylang/#extensions-operations-experimental) is still lacking/buggy.

**Comment by sealmove** [comment]
Fixed by 614a3ea7c9ed5d25c19673864bcbebacfe8bcf07.
